### PR TITLE
Multirow filtered rep

### PR DESCRIPTION
**Pull request by ashudeep-sharma**
This Pull Request tracks the changes for implementing multiValueInsert in filtered Replication side.
I have added some of tests in keyrange_filter_test.go and sqlannotation_test.go. Let me know if some other tests needs to be added.

---
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/2369)

### CODE REVIEW

File: go/vt/binlog/keyrange_filter.go [hunk 1]
```diff
@@ -39,12 +43,32 @@ func KeyRangeFilterFunc(keyrange *topodatapb.KeyRange, sendReply sendTransaction
...
+					}
+					filtered = append(filtered, statement)
+					matched = true
+				} else {
+					multipleQueries, err := splitMultiValueInsertQuery(string(statement.Sql), keyspaceIDS)
+					if err != nil {
+						continue
+					}
+					for rowNum, query := range multipleQueries {
+						if !key.KeyRangeContains(keyrange, keyspaceIDS[rowNum]) {
```
Comment (sougou) [inline go/vt/binlog/keyrange_filter.go]: We may need to bulk up these queries for efficiency. Basically, make splitMultiValueInsertQuery return a single sql that's applicable to the current keyrange.

**Comment by ashudeep-sharma** [comment]
*[go/vt/binlog/keyrange_filter.go, line 59 at r6](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShH-r6-59:-KYz10bgyY9eGk6bFw_h:b3geakq) ([raw file](https://github.com/youtube/vitess/blob/0a1208189fae953b97f2b55d0e76c94ef61f1a8e/go/vt/binlog/keyrange_filter.go#L59)):*
<details><summary><i>Previously, sougou wrote…</i></summary><blockquote>

We may need to bulk up these queries for efficiency. Basically, make splitMultiValueInsertQuery return a single sql that's applicable to the current keyrange.
</blockquote></details>

Done. Incorporated the change in the latest commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2369)*

File: go/vt/binlog/keyrange_filter.go [hunk 2]
```diff
@@ -20,7 +24,7 @@ import (
 func KeyRangeFilterFunc(keyrange *topodatapb.KeyRange, sendReply sendTransactionFunc) sendTransactionFunc {
 	return func(reply *binlogdatapb.BinlogTransaction) error {
 		matched := false
-		filtered := make([]*binlogdatapb.BinlogTransaction_Statement, 0, len(reply.Statements))
+		filtered := []*binlogdatapb.BinlogTransaction_Statement{}
```
Comment (alainjobart) [inline go/vt/binlog/keyrange_filter.go]: Why this change? pre-allocating the filtered array with the approximate size sounds better?

File: go/vt/binlog/keyrange_filter.go [hunk 3]
```diff
@@ -39,12 +43,26 @@ func KeyRangeFilterFunc(keyrange *topodatapb.KeyRange, sendReply sendTransaction
...
-				if !key.KeyRangeContains(keyrange, keyspaceID) {
-					// Skip keyspace ids that don't belong to the destination shard.
-					continue
+				if len(keyspaceIDS) == 1 {
+					if !key.KeyRangeContains(keyrange, keyspaceIDS[0]) {
+						// Skip keyspace ids that don't belong to the destination shard.
+						continue
+					}
+					filtered = append(filtered, statement)
+					matched = true
```
Comment (alainjobart) [inline go/vt/binlog/keyrange_filter.go]: I think we can put a 'continue' here and then the next block can be un-indented, and not in the 'else'?

File: go/vt/binlog/keyrange_filter.go [hunk 4]
```diff
@@ -39,12 +43,26 @@ func KeyRangeFilterFunc(keyrange *topodatapb.KeyRange, sendReply sendTransaction
...
+				} else {
+					query, err := getValidRangeQuery(string(statement.Sql), keyspaceIDS, keyrange)
+					if err != nil || query == "" {
```
Comment (alainjobart) [inline go/vt/binlog/keyrange_filter.go]: The 'err != nil' case should probably log something?

File: go/vt/binlog/keyrange_filter.go [hunk 5]
```diff
@@ -60,6 +78,56 @@ func KeyRangeFilterFunc(keyrange *topodatapb.KeyRange, sendReply sendTransaction
...
+				queryBuf.Myprintf("%s%v", prefix, val)
+				prefix = ", "
+			}
+		}
+		if prefix == "" {
+			return "", nil
+		}
+		queryBuf.Myprintf("%v%s", ins.OnDup, trailingComments)
+		query := queryBuf.String()
+		return query, nil
```
Comment (alainjobart) [inline go/vt/binlog/keyrange_filter.go]: In this section, could we just build a new 'rows' value, and then re-print the SQL statement as is with it, instead of re-generating it all? Would that be simpler?

File: go/vt/sqlannotation/sqlannotation.go [hunk 6]
```diff
@@ -67,51 +68,54 @@ func IsDML(sql string) bool {
...
 // error value.
-func ExtractKeySpaceID(sql string) (keyspaceID []byte, err error) {
-	keyspaceIDString, hasKeySpaceID := extractStringBetween(sql, "/* vtgate:: keyspace_id:", " ")
+func ExtractKeySpaceIDS(sql string) (keyspaceIDs [][]byte, err error) {
+	_, comments := sqlparser.SplitTrailingComments(sql)
+	keyspaceIDString, hasKeySpaceID := extractStringBetween(comments, "/* vtgate:: keyspace_id:", " ")
 	hasUnfriendlyAnnotation := (strings.Index(sql, filteredReplicationUnfriendlyAnnotation) != -1)
-	err = nil
+	ksidStr := strings.Split(keyspaceIDString, ",")
+	keyspaceIDs = make([][]byte, len(ksidStr))
```
Comment (alainjobart) [inline go/vt/sqlannotation/sqlannotation.go]: These two lines can be moved inside the next 'if hasKeySpaceID' block, right?

File: go/vt/sqlannotation/sqlannotation.go [hunk 7]
```diff
@@ -67,51 +68,54 @@ func IsDML(sql string) bool {
...
-// ExtractKeySpaceID parses the annotation of the given statement and tries
+// ExtractKeySpaceIDS parses the annotation of the given statement and tries
 // to extract the keyspace id.
 // If a keyspace-id comment exists 'keyspaceID' is set to the parsed keyspace id
 // and err is set to nil; otherwise, if a filtered-replication-unfriendly comment exists
 // or some other parsing error occured, keyspaceID is set to nil and err is set to a non-nil
 // error value.
-func ExtractKeySpaceID(sql string) (keyspaceID []byte, err error) {
-	keyspaceIDString, hasKeySpaceID := extractStringBetween(sql, "/* vtgate:: keyspace_id:", " ")
+func ExtractKeySpaceIDS(sql string) (keyspaceIDs [][]byte, err error) {
```
Comment (alainjobart) [inline go/vt/sqlannotation/sqlannotation.go]: We use Keyspace as a word, not as two words, so it should be 'Keyspace' not 'KeySpace', in this function name, and below in 'hasKeySpaceID'.

**Comment by ashudeep-sharma** [comment]
*[go/vt/binlog/keyrange_filter.go, line 27 at r8](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShH-r8-27:-KZ5NL49Ngmlmd1BcglI:b-imcx5r) ([raw file](https://github.com/youtube/vitess/blob/1de9e41334c5da8ac5ad8999359c597f456071d7/go/vt/binlog/keyrange_filter.go#L27)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

Why this change? pre-allocating the filtered array with the approximate size sounds better?
</blockquote></details>

This was done initially, since a multivalue insert was being broken to individual statements. However after discussion with Sugu, it came up that let it go as one multiValue insert statement only. I have reverted this change in the latest commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2369)*

**Comment by ashudeep-sharma** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[go/vt/binlog/keyrange_filter.go, line 52 at r8](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShH-r8-52:-KZ5NhdAMw6MVcBeJydu:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/1de9e41334c5da8ac5ad8999359c597f456071d7/go/vt/binlog/keyrange_filter.go#L52)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

I think we can put a 'continue' here and then the next block can be un-indented, and not in the 'else'?
</blockquote></details>

Done.

---

*[go/vt/binlog/keyrange_filter.go, line 55 at r8](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShH-r8-55:-KZ5NjL0P1qHCDduoGJQ:b-saonen) ([raw file](https://github.com/youtube/vitess/blob/1de9e41334c5da8ac5ad8999359c597f456071d7/go/vt/binlog/keyrange_filter.go#L55)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

The 'err != nil' case should probably log something?
</blockquote></details>

Logged the Sql statement as well as the corresponding error.

---

*[go/vt/binlog/keyrange_filter.go, line 124 at r8](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShH-r8-124:-KZ5NqYoT2G_3sq2G6Zi:bnqnt02) ([raw file](https://github.com/youtube/vitess/blob/1de9e41334c5da8ac5ad8999359c597f456071d7/go/vt/binlog/keyrange_filter.go#L124)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

In this section, could we just build a new 'rows' value, and then re-print the SQL statement as is with it, instead of re-generating it all? Would that be simpler?
</blockquote></details>

This sounds much better. I have incorporated this change. Thanks.

---

*[go/vt/sqlannotation/sqlannotation.go, line 77 at r8](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShI-r8-77:-KZ5NvlPB3XkC0gYXnst:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/1de9e41334c5da8ac5ad8999359c597f456071d7/go/vt/sqlannotation/sqlannotation.go#L77)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

We use Keyspace as a word, not as two words, so it should be 'Keyspace' not 'KeySpace', in this function name, and below in 'hasKeySpaceID'.
</blockquote></details>

Done.

---

*[go/vt/sqlannotation/sqlannotation.go, line 82 at r8](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShI-r8-82:-KZ5NwkswWqaAbjXtc6P:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/1de9e41334c5da8ac5ad8999359c597f456071d7/go/vt/sqlannotation/sqlannotation.go#L82)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

These two lines can be moved inside the next 'if hasKeySpaceID' block, right?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2369)*

**Comment by ashudeep-sharma** [comment]
I have incorporated all the review comments. I will be adding a test-case for this in resharding.py

File: go/vt/binlog/keyrange_filter.go [hunk 8]
```diff
@@ -60,6 +82,52 @@ func KeyRangeFilterFunc(keyrange *topodatapb.KeyRange, sendReply sendTransaction
...
+		return "", errors.New("unsupported construct ")
+	}
+}
+
+func generateSingleInsertQuery(ins *sqlparser.Insert, keyspaceIDs [][]byte, trailingComments string, keyrange *topodatapb.KeyRange) (query string, err error) {
+	switch rows := ins.Rows.(type) {
+	case *sqlparser.Select, *sqlparser.Union:
+		return "", errors.New("unsupported: insert into select")
+	case sqlparser.Values:
+		values := sqlparser.Values{}
```
Comment (sougou) [inline go/vt/binlog/keyrange_filter.go]: Saves allocation if nothing matches: `var values sqlparser.Values`

**Comment by sougou** [comment]
For the test case, we discussed 0+1 and 1+2. It just occurred to me that 0+1 is the normal single row insert case. What we really want is 0+2 and 1+2.

File: go/vt/sqlannotation/sqlannotation.go [hunk 9]
```diff
@@ -67,51 +68,54 @@ func IsDML(sql string) bool {
...
+	keyspaceIDString, hasKeyspaceID := extractStringBetween(comments, "/* vtgate:: keyspace_id:", " ")
 	hasUnfriendlyAnnotation := (strings.Index(sql, filteredReplicationUnfriendlyAnnotation) != -1)
-	err = nil
-	if hasKeySpaceID {
+	if hasKeyspaceID {
+		ksidStr := strings.Split(keyspaceIDString, ",")
+		keyspaceIDs = make([][]byte, len(ksidStr))
 		if hasUnfriendlyAnnotation {
-			keyspaceID = nil
+			keyspaceIDs = nil
```
Comment (alainjobart) [inline go/vt/sqlannotation/sqlannotation.go]: I would just:
return nil, &ExtractKeySpaceIDError{
...
}
here, that avoids a level of indentation.

then the two lines ksidStr := ... and keyspaceIDs = ... can be moved after the check.

File: go/vt/sqlannotation/sqlannotation.go [hunk 10]
```diff
@@ -67,51 +68,54 @@ func IsDML(sql string) bool {
...
-		keyspaceID, err = hex.DecodeString(keyspaceIDString)
-		if err != nil {
-			keyspaceID = nil
+	} else {
+		if hasUnfriendlyAnnotation {
 			err = &ExtractKeySpaceIDError{
-				Kind: ExtractKeySpaceIDParseError,
-				Message: fmt.Sprintf(
-					"Error parsing keyspace id value in statement: %v (%v)", sql, err),
+				Kind:    ExtractKeySpaceIDReplicationUnfriendlyError,
```
Comment (alainjobart) [inline go/vt/sqlannotation/sqlannotation.go]: same here, just return nil, &ExtractKeySpaceIDError{...}

File: go/vt/sqlannotation/sqlannotation.go [hunk 11]
```diff
@@ -67,51 +68,54 @@ func IsDML(sql string) bool {
...
+				Message: fmt.Sprintf("Statement: %v", sql),
+			}
+			keyspaceIDs = nil
+		} else {
+			// No annotations.
+			keyspaceIDs = nil
+			err = &ExtractKeySpaceIDError{
+				Kind:    ExtractKeySpaceIDParseError,
+				Message: fmt.Sprintf("No annotation found in '%v'", sql),
```
Comment (alainjobart) [inline go/vt/sqlannotation/sqlannotation.go]: These two cases always return?

Then again, I would swap the original
if hasKeyspaceID {
...
} else {
...
}

into:

if !hasKeyspaceId {
  if A {
   return ...
  } else {
     return ...
  }
}

then no need to indent.

**Comment by ashudeep-sharma** [comment]
Review status: 0 of 6 files reviewed at latest revision, 12 unresolved discussions.

---

*[go/vt/binlog/keyrange_filter.go, line 109 at r10](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShH-r10-109:-KZNxKP2Qz1YmspvgCJW:bb6un1j) ([raw file](https://github.com/youtube/vitess/blob/1848ee2399682c5c17501ba29f0b5a7c3a7985ca/go/vt/binlog/keyrange_filter.go#L109)):*
<details><summary><i>Previously, sougou wrote…</i></summary><blockquote>

Saves allocation if nothing matches: `var values sqlparser.Values`
</blockquote></details>

Done

---

*[go/vt/binlog/keyrange_filter.go, line 124 at r10](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShH-r10-124:-KZNxMHQpJ5GWVIIln29:bb6un1j) ([raw file](https://github.com/youtube/vitess/blob/1848ee2399682c5c17501ba29f0b5a7c3a7985ca/go/vt/binlog/keyrange_filter.go#L124)):*
<details><summary><i>Previously, sougou wrote…</i></summary><blockquote>

Nit (this saves one allocation):
```
queryBuf.WriteString(trailingComments)
return queryBuf.String(), nil
```
</blockquote></details>

Done

---

*[go/vt/sqlannotation/sqlannotation.go, line 85 at r10](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShI-r10-85:-KZNxPiadaOCE_b3i7vO:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/1848ee2399682c5c17501ba29f0b5a7c3a7985ca/go/vt/sqlannotation/sqlannotation.go#L85)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

I would just:
return nil, &ExtractKeySpaceIDError{
...
}
here, that avoids a level of indentation.

then the two lines ksidStr := ... and keyspaceIDs = ... can be moved after the check.
</blockquote></details>

Done.

---

*[go/vt/sqlannotation/sqlannotation.go, line 107 at r10](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShI-r10-107:-KZNxTyvGfERH8rvWcZ0:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/1848ee2399682c5c17501ba29f0b5a7c3a7985ca/go/vt/sqlannotation/sqlannotation.go#L107)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

same here, just return nil, &ExtractKeySpaceIDError{...}
</blockquote></details>

Done.

---

*[go/vt/sqlannotation/sqlannotation.go, line 116 at r10](https://reviewable.io:443/reviews/youtube/vitess/2369#-KYx6xU4gA65bxl8jShI-r10-116:-KZNxVDKInEFx6sz1X2s:bv0qatk) ([raw file](https://github.com/youtube/vitess/blob/1848ee2399682c5c17501ba29f0b5a7c3a7985ca/go/vt/sqlannotation/sqlannotation.go#L116)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

These two cases always return?

Then again, I would swap the original
if hasKeyspaceID {
...
} else {
...
}

into:

if !hasKeyspaceId {
  if A {
   return ...
  } else {
     return ...
  }
}

then no need to indent.
</blockquote></details>

Done. Changed the logic according to the way you have suggested and also what was recommended by golint :).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2369)*

**Comment by ashudeep-sharma** [comment]
I have made the changes for incorporating the review comments and adding the test for multi_row_insert during re-sharding for both the scenarios that we discussed.
Please take a look.

**Comment by sougou** [comment]
It looks like the resharding test is failing on Travis. Once we figure out and fix the root cause, we should be good to merge.

**Comment by sougou** [comment]
Ok. I've found the problem. You have accidentally changed the file mode of `resharding.py` to 644. You can fix this by running the following commands:
```
git update-index --add --chmod=+x test/resharding.py
git commit -m 'Fixing file mode'
```
Also, gpylint generated some errors that you have to fix: https://gist.github.com/sougou/53b972250fdb6690e011b1bea71ff36f

**Comment by ashudeep-sharma** [comment]
Since pylint doesn't work on OS X, i was making changes on mac and trying it out on ubuntu. It seems like due to this i might have changed the file permissions. I have fixed it in this commit and also the pylint issues. Lets check this version of Travis.

**Comment by sougou** [comment]
I think the permission is still 644. The build is still failing.

**Comment by sougou** [comment]
LGTM

**Comment by sougou** [comment]
vtctld_web looks flaky. I'm going to ignore that failure.
